### PR TITLE
[IOTDB-632]Performance improve for Linear Fill

### PR DESCRIPTION
**Pull request by wshao08**
This PR is a reimplementation for Linear Fill.

Consider the Linear Fill function within range [T-before_range, T+after_range]. 
* Search the "before_value" within [T-before_range, T] in a reverse order.
* Search the "after_value" within [T, T+after_range] using aggregation functions. 

**Performance Test result:**

**IoTDB configuration:**
enable_parameter_adapter=false
tsfile_size_threshold=1M
memtable_size_threshold=1M

**Timeseries and Data preparation:**
```
session.createTimeseries("root.test.d1.s1", TSDataType.INT64, TSEncoding.PLAIN, CompressionType.SNAPPY);
session.createTimeseries("root.test.d1.s2", TSDataType.INT64, TSEncoding.PLAIN, CompressionType.SNAPPY);
session.createTimeseries("root.test.d1.s3", TSDataType.INT64, TSEncoding.PLAIN, CompressionType.SNAPPY);
session.createTimeseries("root.test.d2.s1", TSDataType.INT64, TSEncoding.PLAIN, CompressionType.SNAPPY);
session.createTimeseries("root.test.d2.s2", TSDataType.INT64, TSEncoding.PLAIN, CompressionType.SNAPPY);
session.createTimeseries("root.test.d2.s3", TSDataType.INT64, TSEncoding.PLAIN, CompressionType.SNAPPY);
```
Case 1: 10000 seq files, 1000 unseq files. 1000 points in each TsFile.
Case 2: 1000 seq files, 500 unseq files, 1000 points in each TsFile.

**Query Result:** 

 ```Q1:   select s1, s2, s3 from root.test.d1 where time = T_1 Fill(int64[linear])```  
Timestamp [T_1] is only covered with seq files.

```Q2:    select s1, s2, s3 from root.test.d1 where time = T_2 Fill(int64[linear])``` 
Timestamp [T_2] is covered by both seq files and several unseq files.

Below is one of the typical query performance result.

Case 1: 10000 seqFiles  1000 unseq files, 1000 rows in each file.
Master Version
q1: 1696ms, q2: 2943ms 
New Version
q1: 38ms, q2: 673ms 

Case 2: 1000 seqFiles  100 unseen files, 1000 rows in each file
Master Version:
q1: 201ms, q2: 577ms 
New Version:
q1: 12ms, q2: 72ms

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/query/executor/fill/PreviousFill.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  public TimeValuePair getFillResult()
+      throws IOException, QueryProcessException, StorageEngineException {
+    QueryDataSource dataSource =
+        QueryResourceManager.getInstance().getQueryDataSource(seriesPath, context, timeFilter);
+    // update filter by TTL
+    timeFilter = dataSource.updateFilterUsingTTL(timeFilter);
+    LastPointReader lastReader = new LastPointReader(
+        seriesPath, dataType, allSensors, context, dataSource, queryTime, timeFilter);
+
+    return lastReader.getTimeseriesLastPointXXX();
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/executor/fill/PreviousFill.java:107]: XXX?

Comment (wshao08) [reply server/src/main/java/org/apache/iotdb/db/query/executor/fill/PreviousFill.java]: Oh, that's a temporary name forgot to correct.. Fixed

File: server/src/main/java/org/apache/iotdb/db/query/executor/fill/LinearFill.java [hunk 2]
```diff
@@ -83,57 +92,80 @@ public IFill copy() {
...
+        QueryResourceManager.getInstance().getQueryDataSource(seriesPath, context, beforeFilter);
+    LastPointReader lastReader =
+        new LastPointReader(seriesPath, dataType, allSensors, context, dataSource, queryTime, beforeFilter);
+
+    TimeValuePair beforePair = lastReader.readLastPoint();
+    TimeValuePair afterPair = calculatFirstPointAfterQueryTime();
 
     // no before data or has data on the query timestamp
-    if (beforePair == null || beforePair.getTimestamp() == queryTime) {
+    if (beforePair.getValue() == null || beforePair.getTimestamp() == queryTime) {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/executor/fill/LinearFill.java:128]: add the LinearFill in UserGuide, explain how we fill when beforePair or afterPair is null

File: docs/zh/SystemDesign/5-DataQuery/9-FillFunction.md [hunk 3]
```diff
@@ -146,3 +146,14 @@ public TimeValuePair getFillResult() throws IOException {
     return lastPointResult;
 }
 ```
+
+# Linear 填充
+
+对于T时间的 Linear Fill 线性填充值是由该时间序列的两个相关值做线性拟合得到的：T之前的最近时间戳对应的值，T之后的最近时间戳对应的值。
```
[the remark this answers is not in the document]
Comment (wshao08) [reply docs/zh/SystemDesign/5-DataQuery/9-FillFunction.md]: Fixed, thx

File: server/src/main/java/org/apache/iotdb/db/query/executor/fill/LinearFill.java [hunk 4]
```diff
@@ -83,57 +89,79 @@ public IFill copy() {
...
     return average(beforePair, afterPair);
   }
 
+  private TimeValuePair calculateFirstPointAfterQueryTime()
+      throws IOException, StorageEngineException, QueryProcessException {
+    TimeValuePair result = new TimeValuePair(0, null);
+
+    List<AggregateResult> aggregateResultList = new ArrayList<>();
+    AggregateResult minTimeResult =
+        AggregateResultFactory.getAggrResultByName(AggregationType.MIN_TIME.toString(), dataType);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/executor/fill/LinearFill.java:145]:
```suggestion
    AggregateResult minTimeResult = new MinTimeAggrResult(dataType);
```

File: server/src/main/java/org/apache/iotdb/db/query/executor/fill/LinearFill.java [hunk 5]
```diff
@@ -83,57 +89,79 @@ public IFill copy() {
...
+    AggregateResult firstValueResult =
+        AggregateResultFactory.getAggrResultByName(
+            AggregationType.FIRST_VALUE.toString(), dataType);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/executor/fill/LinearFill.java:148]:
```suggestion
    AggregateResult firstValueResult = new FirstValueAggrResult(dataType);
```

**Pull request merged**